src/serializer/Emitter.js

```diff
@@ -268,9 +269,22 @@ export class Emitter {
     invariant(this._activeValues.has(value));
     return condition ? value : undefined;
   }
-  emitAfterWaiting(delayReason: void | Value | SerializedBody, dependencies: Array<Value>, func: () => void) {
+  emitAfterWaiting(
+    delayReason: void | Value | SerializedBody,
+    dependencies: Array<Value>,
+    func: () => void,
+    targetBody: ?SerializedBody
+  ) {
     if (!delayReason) {
-      func();
+      if (targetBody == null || targetBody === this._body) {
+        // Emit into current body.
+        func();
+      } else {
+        invariant(!this._isGeneratorBody(targetBody));
+        const oldBody = this.beginEmitting(targetBody.type, targetBody);
+        func();
+        this.endEmitting(targetBody.type, oldBody);
+      }
     } else if (delayReason instanceof Value) {
```

Comment:
Why is it okay to ignore `targetBody` starting from here on? Should there be an invariant that if `!!delayReason`, then it must be that `targetBody == null`?

Reply:
Good point. Yes, targetBody is only supposed to be used while delayReason is undefined branch. I will add this invariant.